src/Npgsql/Internal/TypeHandling/TypeHandlerResolverFactory.cs

```diff
@@ -25,14 +26,18 @@ static class TypeHandlerResolverFactoryExtensions
 
 public class TypeMappingInfo
 {
-    public TypeMappingInfo(NpgsqlDbType? npgsqlDbType, DbType dbType, string? dataTypeName, Type clrType)
-        => (NpgsqlDbType, DbType, DataTypeName, ClrTypes) = (npgsqlDbType, dbType, dataTypeName, new[] { clrType });
+    public TypeMappingInfo(NpgsqlDbType? npgsqlDbType, string? dataTypeName, Type clrType)
+        => (NpgsqlDbType, DataTypeName, ClrTypes) = (npgsqlDbType, dataTypeName, new[] { clrType });
 
-    public TypeMappingInfo(NpgsqlDbType? npgsqlDbType, DbType dbType, string? dataTypeName, params Type[] clrTypes)
-        => (NpgsqlDbType, DbType, DataTypeName, ClrTypes) = (npgsqlDbType, dbType, dataTypeName, clrTypes);
+    public TypeMappingInfo(NpgsqlDbType? npgsqlDbType, string? dataTypeName, params Type[] clrTypes)
+        => (NpgsqlDbType, DataTypeName, ClrTypes) = (npgsqlDbType, dataTypeName, clrTypes);
 
     public NpgsqlDbType? NpgsqlDbType { get; }
-    public DbType DbType { get; }
+    DbType? dbType;
+    public DbType DbType
+        => dbType ??= NpgsqlDbType is null ? DbType.Object : GlobalTypeMapper.NpgsqlDbTypeToDbType(NpgsqlDbType.Value);
     public string? DataTypeName { get; }
     public Type[] ClrTypes { get; }
+
+    internal void Reset() => dbType = null;
```

Comment:
Honestly, I was just thinking of doing the NpgsqlDbTypeToDbType lookup in NpsgqlParameter.DbType without caching it... The whole thing isn't very perf-sensitive. But am OK with this too.

Reply:
Ah...well, let's leave it as is for now, and if you fell it, you can rewrite it later)

Reply:
Nah, it's OK - no need to rewrite :)